### PR TITLE
Test snapshot compatibility between versions also within same cluster

### PR DESCRIPTION
**Pull request by marregui**
The case where there is version diversity within a cluster is missing.
Just wanted to check this in to get feedback on whether I am heading in the correct direction.

### CODE REVIEW

**Comment by marregui** [comment]
retest this please

File: tests/bwc/test_upgrade.py [hunk 1]
```diff
@@ -446,3 +449,87 @@ def start_cluster_and_alter_tables(self, version_def, nodes):
...
+        '''
+
+    CREATE_SNAPSHOT_TPT = "CREATE SNAPSHOT r1.s{} ALL WITH (wait_for_completion = true)"
+
+    RESTORE_SNAPSHOT_TPT = "RESTORE SNAPSHOT r1.s{} ALL WITH (wait_for_completion = true)"
+
+    DROP_DOC_TABLE = 'DROP TABLE t1'
+
+    SUPPORTED_VERSION_TRANSITIONS = (
+        ('3.3.x', 5, 5),
```
Comment (mfussenegger) [inline tests/bwc/test_upgrade.py]: What's the idea behind `idx_start` and `idx_end` ? Why is this necessary?

Comment (marregui) [reply tests/bwc/test_upgrade.py]: I thought that the version path could go between 3.3.idx_start .. 3.3.idx_end, by making both equal I just take the final revision, but there is always the possibility to run incrementally.

**Comment by marregui** [comment]
I am not sure how to fix the build

**Comment by mfussenegger** [comment]
retest this please

File: src/crate/qa/tests.py [hunk 2]
```diff
@@ -191,6 +191,25 @@ def _new_cluster(self, version, num_nodes, settings=None, env=None):
             nodes.append(self._new_node(version, s, env)[0])
         return CrateCluster(nodes)
 
+    def _new_heterogeneous_cluster(self, versions, settings=None, env=None):
+        self.assertTrue(hasattr(self, '_new_node'))
+        num_nodes = len(versions)
+        settings = settings or {}
+        for port in ['transport.tcp.port', 'http.port', 'psql.port']:
+            self.assertFalse(port in settings)
```
Comment (mfussenegger) [inline src/crate/qa/tests.py]: Could use [assertNotIn](https://docs.python.org/3/library/unittest.html?highlight=unittest#unittest.TestCase.assertIn) which gives better error messages.

File: tests/bwc/test_upgrade.py [hunk 3]
```diff
@@ -531,5 +520,56 @@ def test_snapshot_compatibility(self):
                 cluster.stop()
                 prev_version = version
                 num_snapshot += 1
-            shutil.rmtree(cluster_settings['path.data'], ignore_errors=True)
+            shutil.rmtree(path_data, ignore_errors=True)
         self._process_on_stop()
+
+class SnapshotHeterogeneousNodesCompatibilityTest(SnapshotCompatibilityTest):
+
+    SUPPORTED_VERSION_TRANSITIONS = ['4.0.x'.replace('x', str(i)) for i in range(0, 9)]
```
Comment (mfussenegger) [inline tests/bwc/test_upgrade.py]: I'd keep it simple and nail it to `4.0.8` plus `4.0` and remove the `.x` replacement + randomization logic.

File: tests/bwc/test_upgrade.py [hunk 4]
```diff
@@ -446,3 +456,120 @@ def start_cluster_and_alter_tables(self, version_def, nodes):
...
+                cluster.stop()
+                prev_version = version
+                num_snapshot += 1
+            shutil.rmtree(path_data, ignore_errors=True)
+        self._process_on_stop()
+
+
+class SnapshotHeterogeneousNodesCompatibilityTest(SnapshotCompatibilityTest):
+
+    SUPPORTED_VERSION_TRANSITIONS = ('4.0', '4.0.8')
```
Comment (mfussenegger) [inline tests/bwc/test_upgrade.py]: Not sure if TRANSITIONS fits here. The cluster stays on the same versions?

File: src/crate/qa/tests.py [hunk 5]
```diff
@@ -173,11 +173,11 @@ def _unicast_hosts(self, num, transport_port=4300):
             for x in range(num)
         ])
 
-    def _new_cluster(self, version, num_nodes, settings=None, env=None):
+    def _new_cluster(self, version, num_nodes, crate_root=None, settings=None, env=None):
```
Comment (mfussenegger) [inline src/crate/qa/tests.py]: It looks like `crate_root` is `None` on all call sites. Do we really need this new parameter?

File: src/crate/qa/tests.py [hunk 6]
```diff
@@ -188,13 +188,34 @@ def _new_cluster(self, version, num_nodes, settings=None, env=None):
         nodes = []
         for id in range(num_nodes):
             s['node.name'] = s['cluster.name'] + '-' + str(id)
-            nodes.append(self._new_node(version, s, env)[0])
+            nodes.append(self._new_node(version, crate_root, s, env)[0])
+        return CrateCluster(nodes)
+
+    def _new_heterogeneous_cluster(self, available_versions, num_nodes, crate_root=None, settings=None, env=None):
```
Comment (mfussenegger) [inline src/crate/qa/tests.py]: `crate_root` and `env` are never used. Can we also remove the params here?

Comment (mfussenegger) [reply src/crate/qa/tests.py]: And maybe rename `available_versions` to `versions` or so. Not sure what `available` adds here. Having a `unavailable_versions` or something like that wouldn't make sense.

File: tests/bwc/test_upgrade.py [hunk 7]
```diff
@@ -446,3 +456,118 @@ def start_cluster_and_alter_tables(self, version_def, nodes):
...
+                    c.execute(self.DROP_DOC_TABLE)
+                self._process_on_stop()
+                prev_version = version
+                num_snapshot += 1
+            shutil.rmtree(path_data, ignore_errors=True)
+
+
+class SnapshotHeterogeneousNodesCompatibilityTest(SnapshotCompatibilityTest):
+
+    SUPPORTED_VERSION = ('4.0', '4.0.8')
```
Comment (mfussenegger) [inline tests/bwc/test_upgrade.py]:
```suggestion
    VERSIONS = ('4.0', '4.0.8')
```

Other versions would also be supported, we just don't test them.

Comment (mfussenegger) [reply tests/bwc/test_upgrade.py]: I think either the commit message or inline here as comment on the `VERSIONS` attribute we should clarify **why** we're testing these versions specificly.

File: tests/bwc/test_upgrade.py [hunk 8]
```diff
@@ -446,3 +456,121 @@ def start_cluster_and_alter_tables(self, version_def, nodes):
...
+                    c.execute(self.DROP_DOC_TABLE)
+                self._process_on_stop()
+                prev_version = version
+                num_snapshot += 1
+            shutil.rmtree(path_data, ignore_errors=True)
+
+
+class SnapshotHeterogeneousNodesCompatibilityTest(SnapshotCompatibilityTest):
+
+    # Patched nodes must be compatible
```
Comment (mfussenegger) [inline tests/bwc/test_upgrade.py]:
```suggestion
    # 4.0.9 contains a snapshotting fix that is version dependent.
```
